v4/newrelic/segments_test.go

```diff
@@ -374,3 +379,246 @@ func TestSegmentsOnEmptyTransaction(t *testing.T) {
+	app := newTestApp(t)
+	txn := app.StartTransaction("transaction")
+	seg := &DatastoreSegment{
+		StartTime: txn.StartSegmentNow(),
+	}
+	seg.End()
+	txn.End()
+
+	txnID := getSpanID(txn.rootSpan.Span)
+	segParentID := getParentID(seg.StartTime.Span)
+
+	if segParentID != txnID {
+		t.Errorf("seg is not a child of txn: segParentID=%s, txnID=%s",
+			segParentID, txnID)
+	}
+	if !spanHasEnded(txn.rootSpan.Span) {
+		t.Error("txn root span wasn't ended")
+	}
+	if !spanHasEnded(seg.StartTime.Span) {
+		t.Error("seg wasn't ended")
+	}
+}
+
+func TestDatastoreSegmentNaming(t *testing.T) {
+	testcases := []struct {
+		seg  *DatastoreSegment
+		name string
+	}{
+		{
+			seg: &DatastoreSegment{
+				Product:            DatastorePostgres,
+				Collection:         "collection",
+				Operation:          "operation",
+				ParameterizedQuery: "parametrized query",
+				QueryParameters: map[string]interface{}{
+					"query": "param",
+				},
+				Host:         "host",
+				PortPathOrID: "port",
+				DatabaseName: "dbname",
+			},
+			name: "parametrized query",
+		},
+		{
+			seg: &DatastoreSegment{
+				Product:            DatastorePostgres,
+				Collection:         "collection",
+				Operation:          "operation",
+				ParameterizedQuery: "",
+				QueryParameters: map[string]interface{}{
+					"query": "param",
+				},
+				Host:         "host",
+				PortPathOrID: "port",
+				DatabaseName: "dbname",
+			},
+			name: "'operation' on 'collection' using 'Postgres'",
+		},
+		{
+			seg: &DatastoreSegment{
+				Product:            DatastorePostgres,
+				Collection:         "",
+				Operation:          "operation",
+				ParameterizedQuery: "",
+				QueryParameters: map[string]interface{}{
+					"query": "param",
+				},
+				Host:         "host",
+				PortPathOrID: "port",
+				DatabaseName: "dbname",
+			},
+			name: "'operation' on 'unknown' using 'Postgres'",
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			if name := tc.seg.name(); name != tc.name {
+				t.Errorf(`incorrect name: actual="%s" expected="%s"`, name, tc.name)
+			}
+		})
+	}
+}
+
+func TestParentingExternalSegment(t *testing.T) {
+	app := newTestApp(t)
+	txn := app.StartTransaction("transaction")
+	seg := &ExternalSegment{
+		StartTime: txn.StartSegmentNow(),
+	}
+	seg.End()
+	txn.End()
+
+	txnID := getSpanID(txn.rootSpan.Span)
+	segParentID := getParentID(seg.StartTime.Span)
+
+	if segParentID != txnID {
+		t.Errorf("seg is not a child of txn: segParentID=%s, txnID=%s",
+			segParentID, txnID)
+	}
+	if !spanHasEnded(txn.rootSpan.Span) {
+		t.Error("txn root span wasn't ended")
+	}
+	if !spanHasEnded(seg.StartTime.Span) {
+		t.Error("seg wasn't ended")
+	}
+}
+
+func TestExternalSegmentNaming(t *testing.T) {
+	testcases := []struct {
+		seg  *ExternalSegment
+		name string
+	}{
+		{
+			seg:  &ExternalSegment{},
+			name: "unknown: unknown",
+		},
+		{
+			seg: &ExternalSegment{
+				Host: "myhost:1234",
+			},
+			name: "unknown: myhost:1234",
```

Comment:
Maybe we should add the value of `Library` at the beginning? So we'd get something like `HTTP GET www.newrelic.com`. That might be especially useful for non-HTTP external segments. That's also close to the semantic convention for HTTP spans, which suggest the minimal solution `"HTTP {METHOD_NAME}"`.